Iio regmap fw

tachicialex opened the pull request:
This is an implementation of the iio-regmap driver. It parses a firmware file which stores register operations in a .csv format and runs them using the linux regmap framework over I2C/SPI.
Implemented register operations are:
- READ
- READ_MASK
- WAIT_MASK
- WAIT_MS
- WRITE
- WRITE_MASK

commodo commented on drivers/iio/regmap/iio-regmap-i2c.c:
```diff
@@ -11,21 +11,30 @@
-	regmap = devm_regmap_init_i2c(client, &iio_regmap_config);
-	if (IS_ERR(regmap)) {
+	regmap = devm_regmap_init_i2c(client, regmap_cfg);
+	if (!regmap) {
```
this looks incorrect;
see docstring:
```
/**
 * devm_regmap_init_i2c() - Initialise managed register map
 *
 * @i2c: Device that will be interacted with
 * @config: Configuration for register map
 *
 * The return value will be an ERR_PTR() on error or a valid pointer
 * to a struct regmap.  The regmap will be automatically freed by the
 * device management code.
 */
```

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -19,6 +20,31 @@ struct iio_regmap {
+	u32 val_bits;
+	int ret;
+
+	ret = device_property_read_u32(dev, "reg_bits", &reg_bits);
```
`reg_bits` -> `reg-bits`?

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -19,6 +20,31 @@ struct iio_regmap {
+		return -EINVAL;
+	}
+
+	ret = device_property_read_u32(dev, "val_bits", &val_bits);
```
`val_bits` -> `val-bits`?

commodo commented on drivers/iio/regmap/iio-regmap-spi.c:
```diff
@@ -22,6 +22,16 @@ static int iio_regmap_spi_probe(struct spi_device *spi)
 	const struct spi_device_id *id = spi_get_device_id(spi);
```
do you need to remove
```
static const struct regmap_config iio_regmap_spi_regmap_config = {
};
```
?

commodo commented on drivers/iio/regmap/iio-regmap-spi.c:
```diff
@@ -22,6 +22,16 @@ static int iio_regmap_spi_probe(struct spi_device *spi)
+	if (ret < 0)
+		return ret;
 
 	regmap = devm_regmap_init_spi(spi, &iio_regmap_spi_regmap_config);
```
pass `regmap_cfg` here instead of `&iio_regmap_spi_regmap_config`?

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -45,11 +46,54 @@ int iio_regmap_read_config(struct device *dev, struct regmap_config *regmap_cfg)
+		dev_err(dev, "Firmware data not loaded.");
+		return ERR_PTR(-EINVAL);
+	}
+	memcpy(reg_ops, data, (firmware->size) * sizeof(char));
```
why not do the parsing here?
i.e. calling `parse_register_ops(firmware->data)` here?

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -39,7 +39,7 @@
 
 #define REG_OP_SIZE		20
-#define WAIT_POLL_TIME_US	1000
+#define WAIT_POLL_NR_TIMES	10
```
this looks like a fix commit;
you can probably move this into the commit that adds this initially

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -341,6 +341,143 @@ static int parse_register_ops(struct device *dev, const char *fw_reg_ops,
+		dev_err(dev, "regmap_write failed, addr: %x", op->addr);
+		return ret;
+	}
+	return 0;
```
this looks like it could be simplified with
```
	if (ret < 0)
		dev_err(dev, "regmap_write failed, addr: %x", op->addr);

	return ret;
```

also, maybe print the value of `ret`

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -341,6 +341,143 @@ static int parse_register_ops(struct device *dev, const char *fw_reg_ops,
+	register_value |= op->val;
+
+	ret = regmap_write(regmap, op->addr, op->val);
+	if (ret < 0) {
```
this looks like it could be simplified with
```
	if (ret < 0)
		dev_err(dev, "regmap_write failed, addr: %x", op->addr);

	return ret;
```

also, maybe print the value of `ret`

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -341,6 +341,143 @@ static int parse_register_ops(struct device *dev, const char *fw_reg_ops,
+			dev_err(dev,
+				"Invalid reg [%x] value:[%x], expected[%x]",
+				op->addr, value, op->val);
+			return -1;
```
-ETIMEDOUT ?  [i don't recall the correct name, but something like that]

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -88,6 +143,229 @@ static const char *read_firmware(struct device *dev)
+{
+	int ret;
+
+	ret = sscanf(reg_ops, ",,,%u,%u", &op->time, &op->dbg);
```
for all these `sscanf` cases, where you need to ignore something, maybe try to check `sscanf(reg_ops, "%i,%i,,%u,%u", &op->time, &op->dbg);`

not sure if it's `%i` or something else, but it would be useful to ignore whatever is in there;
otherwise users add whitespaces and ask use why things don't work, because they can't keep track of their whitespaces

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -88,6 +143,229 @@ static const char *read_firmware(struct device *dev)
+{
+	int ret;
+
+	ret = sscanf(reg_ops, "%x,%x,%x,%u,%u", &op->addr, &op->mask, &op->val,
```
is there a `sscanf` specifier that can handle all int formats?
something that could handle  `0xabcd` [in 16 base] or in base 10 ?

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -88,6 +143,229 @@ static const char *read_firmware(struct device *dev)
+ * finally run each command.
+ */
+static int interpret_register_ops(struct device *dev, struct regmap *regmap,
+				  const char *fw_reg_ops)
```
the name `const char *fw_reg_ops)` is a bit confusing
it's still raw data; maybe something ilke  `const char *text`  or  `const char *input`

`fw_reg_ops` kind of suggests they are already parsed

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -88,6 +143,229 @@ static const char *read_firmware(struct device *dev)
+		memcpy(parsed_cmd, fw_reg_ops, op_size);
+		fw_reg_ops = reg_op_end + 1;
+
+		op_code = regmap_cmd_to_opcode(parsed_cmd);
```
why not interrupt parsing altogether here?
if we get an invalid value here, maybe just stop?

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -13,6 +38,36 @@
 
 #include "iio-regmap.h"
 
+#define REG_OP_SIZE		20
```
maybe name this `REG_OP_STRLEN`
`REG_OP_SIZE` is a bit vague; and can refer to bit-size, or byte-size of a value

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -341,6 +341,143 @@ static int parse_register_ops(struct device *dev, const char *fw_reg_ops,
+			ret = run_wait_mask_op(dev, regmap, reg_op);
+			break;
+		case IIO_REGMAP_WAIT_MS:
+			mdelay(reg_op->time);
```
```
if (reg_op->time);
    mdelay(reg_op->time);
```

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -341,6 +341,143 @@ static int parse_register_ops(struct device *dev, const char *fw_reg_ops,
+			dev_err(dev, "Invalid op code: %d.", reg_op->op);
+			return -EINVAL;
+		}
+	}
```
maybe move
```
if (ret)
    return ret
```

inside the loop here; and remove the one below;
if an op fails,  maybe we want to stop the whole thing

tachicialex commented:
Changelog V2:
- Squashed in one commit the firmware reading and _parse_integer
- Fixed devm_regmap_init_i2c error checking
- Register commands are now case insensitive and may have white spaces inside print_hex_dump
- Register commands can contain base 8, base 16 or base 10 numbers
- Print returned code in case of error when running regmap functions

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -301,6 +302,141 @@ static int parse_register_ops(struct device *dev, const char *input_fw,
+	ret = regmap_write(regmap, op->addr, op->val);
+	if (ret < 0)
+		dev_err(dev, "regmap_write failed, addr: %x, code: %d",
+		op->addr, ret);
```
nitpick: indent this 1 level

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -301,6 +302,141 @@ static int parse_register_ops(struct device *dev, const char *input_fw,
+	wait_time = op->time % WAIT_POLL_NR_TIMES;
+	if (wait_time)
+		mdelay(wait_time);
+	wait_time = op->time / WAIT_POLL_NR_TIMES;
```
if `wait_time` is zero, maybe then just exit here; no need to do the whole loop

commodo replied:
disregard my previous comment;
i just noticed that this is `run_wait_mask_op()` ;

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -301,6 +302,141 @@ static int parse_register_ops(struct device *dev, const char *input_fw,
+	int ret;
+	unsigned int value;
+	unsigned int wait_time;
+	unsigned int wait_nr_times = WAIT_POLL_NR_TIMES;
```
this isn't important here; just a note only;
some upstream people don't are about variable length ordering; some do; there is no explicit rule anywhere, just what maintainers complain about;

the preferred way to order variables, seems to be "reverse-christmas-tree"
so, this would become:
```
	unsigned int wait_nr_times = WAIT_POLL_NR_TIMES;
	unsigned int wait_time;
	unsigned int value;
	int ret;
```

but, as i said: this is only a note; when upstreaming to some subsystems (let's say netdev), they might care;
iio doesn't care that much

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -301,6 +302,141 @@ static int parse_register_ops(struct device *dev, const char *input_fw,
+	ret = regmap_read(regmap, op->addr, &value);
+	if (ret < 0) {
+		dev_err(dev, "regmap_read failed, addr: %x, code: %d",
+		reop->addr, ret);
```
nitpick: indent 1 level

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -301,6 +302,141 @@ static int parse_register_ops(struct device *dev, const char *input_fw,
+	}
+
+	register_value &= ~op->mask;
+	register_value |= op->val;
```
maybe do `register_value |= (op->val & op->mask);`
this is to ensure against any potential issues with values not respecting the mask value

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -394,8 +530,9 @@ int iio_regmap_probe(struct device *dev, struct regmap *regmap,
-	return ret;
+	return iio_run_regmap_ops(dev, regmap, register_ops, nr_ops);
 }
+
```
nitpick: extra line; it's not required

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -45,11 +100,273 @@ int iio_regmap_read_config(struct device *dev, struct regmap_config *regmap_cfg)
+	for (c = fw_reg_ops; *c != '\0'; c++)
+		if (*c == '\n')
+			new_lines++;
+	reg_ops = devm_kzalloc(dev, new_lines * sizeof(*reg_ops), GFP_KERNEL);
```
what happens is `new_lines` is zero here?

commodo replied:
for-loop could use some brackets; the block inside the for-loop is more than 1 instruction
so:
```
	new_lines = 0;
	for (c = fw_reg_ops; *c != '\0'; c++) {
		if (*c == '\n')
			new_lines++;
	}
```

commodo commented on drivers/iio/regmap/iio-regmap-i2c.c:
```diff
@@ -11,18 +11,27 @@
+	if (!regmap_cfg)
+		return -ENOMEM;
+
+	ret = iio_regmap_read_config(&client->dev, regmap_cfg);
```
why not move the allocation of `regmap_cfg` `inside iio_regmap_read_config()` ?
then, if an error occurs you can return `ERR_PTR(ret)` and check wiht `IS_ERR(regmap_cfg)`

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -5,13 +5,68 @@
+	unsigned int		val;
+	unsigned int		time;
+	unsigned int		dbg;
+};
```
maybe a static op table would be an idea
something like
```
static const struct iio_regmap_op  iio_regmap_op_required_fields [] = {
	[IIO_REGMAP_READ_MASK] = { .addr = 1, .mask = 1, etc}
       ........
};
```

This would unify the parsing function, as you can use a single parse function for all ops.
Each op would define it's required fields, and the parse function would error out if the field is not defined;

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -45,11 +100,273 @@ int iio_regmap_read_config(struct device *dev, struct regmap_config *regmap_cfg)
+{
+	int ret;
+
+	ret = sscanf(reg_ops, " %i , , , , %i ", &op->addr, &op->dbg);
```
related to https://github.com/analogdevicesinc/linux/pull/670/commits/a5cd21850153d651c416fb62d77ac1d0243a25b4#r350079707   the parse function would need to use `strtok()` to go over each comma, to identify the position of each field

commodo replied:
i mean strtok() + strtoull()

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -45,11 +100,273 @@ int iio_regmap_read_config(struct device *dev, struct regmap_config *regmap_cfg)
+			return -EINVAL;
+		}
+
+		input_fw = strchr(input_fw, '\n') + 1;
```
what if `strchr(input_fw, '\n')` returns NULL ?

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -45,11 +100,273 @@ int iio_regmap_read_config(struct device *dev, struct regmap_config *regmap_cfg)
+		return ERR_PTR(-ENOMEM);
+	}
+
+	ret = parse_register_ops(dev, input_fw, reg_ops);
```
maybe also pass `firmware->size` and use it for parsing
that looks like a good indicator of where to stop parsing the string

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -45,11 +100,273 @@ int iio_regmap_read_config(struct device *dev, struct regmap_config *regmap_cfg)
+		return NULL;
+
+	new_lines = 0;
+	for (c = fw_reg_ops; *c != '\0'; c++)
```
curios; is the last character guaranteed to be NULL ?

commodo commented on drivers/iio/regmap/iio-regmap.c:
```diff
@@ -45,11 +100,273 @@ int iio_regmap_read_config(struct device *dev, struct regmap_config *regmap_cfg)
+	int i;
+	int op_code = -1;
+
+	for (i = IIO_REGMAP_READ; i <= IIO_REGMAP_WRITE_MASK; i++) {
```
you can use the `match_string()` helper here;
